### PR TITLE
Bug 1219575 - [browser] fetch favor icon from sync bookmark and history data

### PR DESCRIPTION
**Pull request by danhuang1202**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2969d341c7565ee4c73b910b2f16f114c8e56ef3

File: tv_apps/browser/js/sync/db.js [hunk 1]
```diff
@@ -1190,14 +1190,34 @@ SyncBrowserDB.db = {
...
+      transaction.objectStore(DBOS_BOOKMARKS).index('parentid');
+    var iconsStore = transaction.objectStore(DBOS_ICONS);
 
-    objectStore.openCursor(parentId, 'prev').onsuccess =
+    bookmarksStore.openCursor(parentId, 'prev').onsuccess =
       function onSuccess(e) {
       var cursor = e.target.result;
       if (cursor) {
-        records.push(cursor.value);
+        makeBookmarkProcessor(cursor.value);
```
Comment (weilonge) [inline tv_apps/browser/js/sync/db.js]: makeBookmarkProcessor should return Promise to handle the following tasks - records.push(cursor.value); cursor.continue();

File: tv_apps/browser/js/sync/db.js [hunk 2]
```diff
@@ -1190,14 +1190,34 @@ SyncBrowserDB.db = {
...
+        iconsStore.get(data.bmkUri).onsuccess = function(e) {
+          var icons = e.target.result;
+          if(icons){
+            if(icons.iconUri){
+              data.iconUri = icons.iconUri;
+            } else {
+              data.iconUri = Awesomescreen.DEFAULT_FAVICON;
+            }
+          }
+          records.push(data);
```
Comment (weilonge) [inline tv_apps/browser/js/sync/db.js]: Never do records.push(data); in makeBookmarkProcessor

File: tv_apps/browser/js/sync/db.js [hunk 3]
```diff
@@ -1190,14 +1190,34 @@ SyncBrowserDB.db = {
     var records = [];
     var db = this._db;
 
-    var transaction = db.transaction(DBOS_BOOKMARKS);
-    var objectStore = transaction.objectStore(DBOS_BOOKMARKS).index('parentid');
+    function makeBookmarkProcessor(data) {
```
Comment (weilonge) [inline tv_apps/browser/js/sync/db.js]: change the function name to fetch iconUri

File: tv_apps/browser/js/sync/db.js [hunk 4]
```diff
@@ -1190,14 +1190,34 @@ SyncBrowserDB.db = {
...
+      } else {
+        iconsStore.get(data.bmkUri).onsuccess = function(e) {
+          var icons = e.target.result;
+          if(icons){
+            if(icons.iconUri){
+              data.iconUri = icons.iconUri;
+            } else {
+              data.iconUri = Awesomescreen.DEFAULT_FAVICON;
+            }
+          }
```
Comment (weilonge) [inline tv_apps/browser/js/sync/db.js]: Simplify the login:
```js
if (icons && icons.iconUri) {
    data.iconUri = icons.iconUri;
} else {
    data.iconUri = Awesomescreen.DEFAULT_FAVICON;
}
```

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d478cd0d79562b6c7db8ac93c5dc7e2066b89da3

File: tv_apps/browser/js/sync/db.js [hunk 5]
```diff
@@ -1190,14 +1190,40 @@ SyncBrowserDB.db = {
...
-    objectStore.openCursor(parentId, 'prev').onsuccess =
+    bookmarksStore.openCursor(parentId, 'prev').onsuccess =
       function onSuccess(e) {
       var cursor = e.target.result;
       if (cursor) {
-        records.push(cursor.value);
+        fetchIconUri(cursor.value).then(bookmark => {
+          records.push(bookmark);
+        });
         cursor.continue();
```
Comment (weilonge) [inline tv_apps/browser/js/sync/db.js]: hmm, I suspect it might have a race-condition here, but the code flow seems fine. Let's figure it later.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f47f1a96bf70c9959d2bce88082768ce16f98467

**Pull request merged**